src/main/java/org/broadinstitute/hellbender/tools/copynumber/coverage/denoising/svd/SVDDenoisingUtils.java

```diff
@@ -0,0 +1,482 @@
+
+        final RealMatrix denoisedCopyRatioValues;
+        if (panelOfNormals.getOriginalReadCounts().length == 1) {
+            logger.warn("Only a single sample was used to build the panel of normals, not cannot perform denoising...");
+            denoisedCopyRatioValues = standardizedCopyRatioValues;
+        } else {
+            logger.info("Subtracting projection onto space spanned by eigensamples...");
+            denoisedCopyRatioValues = subtractProjection(standardizedCopyRatioValues, panelOfNormals.getEigensampleVectors(), numEigensamples);
+        }
+
+        logger.info("Sample denoised.");
+
+        //construct the result
+        return new SVDDenoisedCopyRatioResult(
+                new SimpleSampleMetadata(readCounts.getSampleName()),
+                panelOfNormals.getPanelIntervals(),
+                standardizedCopyRatioValues,
+                denoisedCopyRatioValues);
+    }
+
+    /**
+     * Preprocess (i.e., transform to fractional coverage and correct GC bias)
+     * and standardize read counts for samples when no panel of normals is available.
+     * The original {@code readCounts} has dimensions 1 x intervals and is not modified.
+     * If {@code intervalGCContent} is null, GC-bias correction will not be performed
+     */
+    public static RealMatrix preprocessAndStandardizeSample(final double[] readCounts,
+                                                            final double[] intervalGCContent) {
+        Utils.nonNull(readCounts);
+        Utils.validateArg(intervalGCContent == null || readCounts.length == intervalGCContent.length,
+                "Number of intervals for read counts must match those for GC-content annotations.");
+
+        RealMatrix result = new Array2DRowRealMatrix(new double[][]{readCounts});
+
+        //preprocess (transform to fractional coverage, correct GC bias) copy in place
+        logger.info("Preprocessing read counts...");
+        transformToFractionalCoverage(result);
+        performOptionalGCBiasCorrection(result, intervalGCContent);
+        logger.info("Sample read counts preprocessed.");
+
+        //standardize copy in place
+        logger.info("Standardizing read counts...");
+        divideBySampleMedianAndTransformToLog2(result);
+        logger.info("Subtracting sample median...");
+        final double[] sampleLog2Medians = MatrixSummaryUtils.getRowMedians(result);
+        result.walkInOptimizedOrder(new DefaultRealMatrixChangingVisitor() {
+            @Override
+            public double visit(int sampleIndex, int intervalIndex, double value) {
+                return value - sampleLog2Medians[sampleIndex];
+            }
+        });
+        logger.info("Sample read counts standardized.");
+
+        return result;
+    }
+
+    private static PreprocessedStandardizedResult preprocessPanel(final RealMatrix readCounts,
```

Comment:
This method is really long.  Can it be simplified or broken into sub-methods?

Reply:
Unfortunately, refactoring this code so that it is memory efficient (using the filter arrays) and broken up into methods makes it much less readable and not very reusable, in my opinion (I actually wrote the code that way originally).

Reply:
If methods are passed by reference, how can you not even break it up into sub-methods?

Regardless, I am okay with this long method, if you add a comment as to why.

Reply:
The signatures of some (but not all) of the sub-methods must then include both filter arrays, which makes them inconsistent and awkward to reuse.

I'll add some comments about why this code is long and a warning not to split it up or refactor without good reason.